File: 2016/03.cs
namespace _2016
{
    public abstract class _03
    {
        private static readonly string[] Data;

        static _03() => Data = Task.Run(() => Utils.GetInputData(3)).Result
            .Split('\n', StringSplitOptions.RemoveEmptyEntries);

        private static bool ValidTriangle(int[] sides) =>
            sides[0] + sides[1] > sides[2] && sides[1] + sides[2] > sides[0] && sides[2] + sides[0] > sides[1];

        private static int SolvePartOne() =>
            Data.Select(line =>
                    line.Trim().Split([' '], StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToArray())
                .Count(ValidTriangle);

        private static int SolvePartTwo()
        {
            var validTriangleCount = 0;
            var columns = new List<List<int>> { new(), new(), new() };

            for (var row = 0; row < Data.Length; row += 3)
            {
                var firstRow = Data[row].Trim().Split([' '], StringSplitOptions.RemoveEmptyEntries)
                    .Select(int.Parse).ToArray();
                var secondRow = Data[row + 1].Trim().Split([' '], StringSplitOptions.RemoveEmptyEntries)
                    .Select(int.Parse).ToArray();
                var thirdRow = Data[row + 2].Trim().Split([' '], StringSplitOptions.RemoveEmptyEntries)
                    .Select(int.Parse).ToArray();

                for (var col = 0; col < 3; col++)
                {
                    columns[col].Add(firstRow[col]);
                    columns[col].Add(secondRow[col]);
                    columns[col].Add(thirdRow[col]);
                }
            }

            foreach (var col in columns)
            {
                for (var i = 0; i < col.Count; i += 3)
                {
                    var sides = new[] { col[i], col[i + 1], col[i + 2] };
                    if (ValidTriangle(sides)) validTriangleCount++;
                }
            }

            return validTriangleCount;
        }

        public static void Run()
        {
            Console.WriteLine($"Part One: {SolvePartOne()}");
            Console.WriteLine($"Part Two: {SolvePartTwo()}");
        }
    }
}

File: 2018/03.js
import {getInputData} from './utils.js';

const data = (await getInputData(3)).split('\n').filter(x => x !== '');

function processClaims(processClaimHandler) {
	const fabric = new Map();
	for (const claim of data) {
		const [id, , at, size] = claim.split(' ');
		const claimId = Number(id.replace('#', ''));
		const [x, y] = at.replace(':', '').split(',').map(Number);
		const [width, height] = size.split('x').map(Number);
		for (let i = x; i < x + width; i++) for (let j = y; j < y + height; j++) {
			const key = `${i},${j}`;
			processClaimHandler(fabric, key, claimId);
		}
	}
	return fabric;
}

function solvePartOne() {
	let overlapCount = 0;
	processClaims((fabric, key) => {
		fabric.set(key, (fabric.get(key) || 0) + 1);
		if (fabric.get(key) === 2) overlapCount++;
	});
	return overlapCount;
}

function solvePartTwo() {
	const overlappingClaims = new Set();
	const allClaims = new Set();
	processClaims((fabric, key, claimId) => {
		allClaims.add(claimId);
		if (!fabric.has(key)) fabric.set(key, [claimId]);
		else {
			fabric.get(key).forEach(existingClaimId => overlappingClaims.add(existingClaimId));
			overlappingClaims.add(claimId);
			fabric.get(key).push(claimId);
		}
	});
	for (const claimId of allClaims) if (!overlappingClaims.has(claimId)) return claimId;
	return null;
}

console.log(`Part One: ${solvePartOne()}`);
console.log(`Part Two: ${solvePartTwo()}`);